CIScripts/Test/PesterLogger/RemoteLogCollector.ps1

```diff
@@ -1,11 +1,94 @@
+. $PSScriptRoot/../../Common/Aliases.ps1
+. $PSScriptRoot/../../Common/Invoke-NativeCommand.ps1
+
 . $PSScriptRoot/PesterLogger.ps1
 
-function New-LogSource {
+class LogSource {
+    [System.Management.Automation.Runspaces.PSSession] $Session
+
+    [Hashtable] GetContent() {
+        throw "LogSource is an abstract class, use specific log source instead"
+    }
+
+    ClearContent() {}
+}
+
+class FileLogSource : LogSource {
+    [String] $Path
+
+    [Hashtable] GetContent() {
+        $ContentGetterBody = {
+            Param([Parameter(Mandatory = $true)] [string] $From)
+            $Files = Get-ChildItem -Path $From -ErrorAction SilentlyContinue
+            $Logs = @{}
+            if (-not $Files) {
+                $Logs[$From] = "<FILE NOT FOUND>"
+            } else {
+                foreach ($File in $Files) {
+                    $Content = Get-Content -Raw $File
+                    $Logs[$File.FullName] = if ($Content) {
+                        $Content
+                    } else {
+                        "<FILE WAS EMPTY>"
+                    }
+                }
+            }
+            return $Logs
+        }
+        return Invoke-CommandRemoteOrLocal -Func $ContentGetterBody -Session $this.Session -Arguments $this.Path
+    }
+
+    ClearContent() {
+        $LogCleanerBody = {
+            Param([Parameter(Mandatory = $true)] [string] $What)
+            $Files = Get-ChildItem -Path $What -ErrorAction SilentlyContinue
+            foreach ($File in $Files) {
+                try {
+                    Remove-Item $File
+                }
+                catch {
+                    Write-Warning "$File was not removed due to $_"
+                }
+            }
+        }
+        Invoke-CommandRemoteOrLocal -Func $LogCleanerBody -Session $this.Session -Arguments $this.Path
+    }
+}
+
+class ContainerLogSource : LogSource {
+    [String] $Container
+
+    [Hashtable] GetContent() {
+        $Command = Invoke-NativeCommand -Session $this.Session -CaptureOutput -AllowNonZero {
+            docker logs $Using:this.Container
+        }
+        return @{
+            "$( $this.Container ) container logs" = $Command.Output
+        }
+    }
+}
+
+function New-ContainerLogSource {
+    Param([Parameter(Mandatory = $true)] [string[]] $ContainerNames,
+          [Parameter(Mandatory = $false)] [PSSessionT[]] $Sessions)
+
+    return $Sessions | ForEach-Object {
+        $Session = $_
+        $ContainerNames | ForEach-Object {
+            [ContainerLogSource] @{
+                Session = $Session
+                Container = $_
+            }
+        }
+    }
+}
+
+function New-FileLogSource {
     Param([Parameter(Mandatory = $true)] [string] $Path,
           [Parameter(Mandatory = $false)] [PSSessionT[]] $Sessions)
 
     return $Sessions | ForEach-Object {
-        @{
+        [FileLogSource] @{
```

Comment:
Possible refinement: what do you think about writing a constructor?

Reply:
I'm against writing constructors for plain-data classes.

Reply:
Except it has two methods, it's not plain data at all? :confused:

Reply:
(but whatever, it's possible refinement so I approve)

Reply:
You're right. I think I've forgotten that I had refactored it to *have* methods.

Anyway, if the constructor does nothing more than setting fields to arguments, I see no point in creating it. Especially since this class is an implementation detail (though I don't know if you can mark something as private in PS), and the `New-FileLogSource` is the user-facing constructor.

Reply:
We were wondering for calling constructor if e.g. only one remote session is used, and calling `New-*LogSource`*s* when multiple are. But since you pointed out that convention is not to have plurals for cmdlet names, then I think this is a moot point and I agree.